server/controler/login.go

```diff
@@ -0,0 +1,81 @@
+package controler
+
+import (
+	"encoding/json"
+	"net/http"
+	"strings"
+
+	"golang.org/x/crypto/bcrypt"
+)
+
+type loginReq struct {
+	Username string
+	Password string
+}
+
+type loginResponse struct {
+	Token string `json:"token"`
+}
+
+var minLength = 3
+
+func (env *Env) Login(responseWriter http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		// http.Error(responseWriter,
+		// http.StatusText(http.StatusMethodNotAllowed),
+		// http.StatusMethodNotAllowed)
+		responseWriter.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	var loginReq loginReq
+	if err := json.NewDecoder(r.Body).Decode(&loginReq); err != nil {
+		http.Error(responseWriter,
+			http.StatusText(http.StatusBadRequest),
+			http.StatusBadRequest)
+		return
+	}
+
+	user, err := env.store.GetUserByUsername(strings.ToLower(loginReq.Username))
+	if err != nil {
+		http.Error(responseWriter,
+			http.StatusText(http.StatusInternalServerError),
+			http.StatusInternalServerError)
+		return
+	} else if user == nil {
+		http.Error(responseWriter, "User not found", http.StatusNotFound) //not sure if correct status
+		return
+	}
+
+	if err := bcrypt.CompareHashAndPassword([]byte(user.PasswordHash),
+		[]byte(loginReq.Password)); err != nil {
+		http.Error(responseWriter,
+			"Wrong username or password",
```

Comment:
what to do with these cases when message and code differs?

Reply:
As I wrote above:
> The general rule is to return HTTP error code + add a more detailed description in the body of the response if possible.

I think it's a good practice adding a description to the error response.